### ISSUE TITLE
False positive in clippy::unnecessary_literal_unwrap if variable is later used for struct construction

### ISSUE THREAD

**Issue by eric-seppanen** [rust-lang/rust-clippy#11381]
### Summary

`clippy::unnecessary_literal_unwrap` suggests removing the `Some` and `unwrap` in places where that is impossible, because we are using the value later to construct a struct that contains a member that's an `Option`.

### Reproducer

I tried this code:

```rust
#[allow(dead_code)]
pub struct Thing {
    value: Option<u32>,
}

pub fn try_thing() -> Thing {
    let value = Some(42u32);
    assert!(value.unwrap() != 0);

    Thing { value }
}
```

I expected to see this happen:

*nothing.*

Instead, this happened:

```
warning: used `unwrap()` on `Some` value
 --> src/lib.rs:8:13
  |
8 |     assert!(value.unwrap() != 0);
  |             ^^^^^^^^^^^^^^
  |
help: remove the `Some` and `unwrap()`
 --> src/lib.rs:7:17
  |
7 |     let value = Some(42u32);
  |                 ^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_literal_unwrap
  = note: `#[warn(clippy::unnecessary_literal_unwrap)]` on by default
```

Following the "help" advice is not going to work, because `value` needs to be an `Option`.

### Version

```text
rustc 1.72.0-beta.11 (7a3a43a3b 2023-08-18)
binary: rustc
commit-hash: 7a3a43a3b984f0d3259a407b17f0b0066f094d39
commit-date: 2023-08-18
host: x86_64-unknown-linux-gnu
release: 1.72.0-beta.11
LLVM version: 16.0.5
```

**Comment by spearman** [comment]
It seems like maybe clippy doesn't check if the Option is used elsewhere? I have a similar false positive where the Option is used in a function call:
```
use std::cell::RefCell;

fn foo(x: &Option<RefCell<Vec<u32>>>) {
  if let Some(x) = x.as_ref() {
    x.borrow_mut().push(5)
  }
}

fn main() {
  let x = Some(RefCell::new(vec![1u32]));
  foo(&x);
  let x = x.unwrap().into_inner();
  dbg!(x);
}
```

```
warning: used `unwrap()` on `Some` value
  --> src/main.rs:12:11
   |
12 |   let x = x.unwrap().into_inner();
   |           ^^^^^^^^^^
   |
help: remove the `Some` and `unwrap()`
  --> src/main.rs:10:11
   |
10 |   let x = Some(RefCell::new(vec![1u32]));
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_literal_unwrap
   = note: `#[warn(clippy::unnecessary_literal_unwrap)]` on by default
```

```
rustc 1.91.0-nightly (9eb4a2652 2025-08-18)
binary: rustc
commit-hash: 9eb4a2652031ed5ba97c29ef21c79db1645f7883
commit-date: 2025-08-18
host: x86_64-unknown-linux-gnu
release: 1.91.0-nightly
LLVM version: 21.1.0
```
